src/tools/optimization-options.h

```diff
@@ -308,11 +337,25 @@ struct OptimizationOptions : public ToolOptions {
       passRunner.setDebug(true);
     }
     for (auto& pass : passes) {
-      if (pass == DEFAULT_OPT_PASSES) {
+      // We apply the pass's intended opt and shrink levels, if any.
+      auto oldOptimizeLevel = passRunner.options.optimizeLevel;
+      auto oldShrinkLevel = passRunner.options.shrinkLevel;
+      if (pass.optimizeLevel) {
+        passRunner.options.optimizeLevel = *pass.optimizeLevel;
+      }
+      if (pass.shrinkLevel) {
+        passRunner.options.shrinkLevel = *pass.shrinkLevel;
+      }
+
+      if (pass.name == DEFAULT_OPT_PASSES) {
         passRunner.addDefaultOptimizationPasses();
       } else {
-        passRunner.add(pass);
+        passRunner.add(pass.name);
       }
+
+      // Revert back to the default levels, if we changed them.
+      passRunner.options.optimizeLevel = oldOptimizeLevel;
+      passRunner.options.shrinkLevel = oldShrinkLevel;
```

Comment:
Can we skip restoring the old levels now that we have per-pass levels? It would be simpler to  not try to mix some concept of global levels with per-pass levels.

Reply:
I think the global level is unavoidable, see the note in the changelog, since we need `--precompute -O1` to still see the opt level of `1` in precompute. That reads a global opt level.

I suppose we could do 2 passes, first set up the pass names, and while doing so find the default, then go over the passes once more and set their opt levels. But I'm not sure that's simpler.

Reply:
While we're changing how things work, can we make it so that `--precompute -O1` runs precompute with the default settings but `-O1 --precompute` runs it with the `-O1` settings?

Reply:
I considered that, yeah - in a way it makes sense. But it would be a larger breaking change. I think this PR just fixes behavior that is obviously surprising and wrong, while what you suggest is more debatable. I'd suggest leaving that to a separate discussion, but I'm open to the idea.

Reply:
How many passes do have different behaviors according to their "global level"? I like the direction this PR is going and I think this itself LGTM, but I think that `--precompute` can do different things based on its location is confusing. Does clang have this kind of behavior too? (I don't think so, but it's possible that I'm not aware of it)

If there are not too many passes that have different behaviors depending on the global level, can we just make different options for those options? We already have several passes that have two options like `--inlining` and `--inlining-optimizing`.

Reply:
IMO having the `-O*` flags be non-idempotent is already surprising. Given that repeating them is meaningful, I think it's a much simpler mental model if they only affect optimization levels for later passes. Otherwise they still have the kind of retroactive "time travelling" behavior that made the semantics this PR fixes so confusing.

Reply:
> How many passes do have different behaviors according to their "global level"?

Looks like around 8, currently.

Reply:
Hmm, yeah, maybe we should add options for each of those. Like `--precompute-3` for opt level 3, or something nicer? Anyhow, yes, we can leave that for follow-up discussions. I feel that isn't urgent, but this PR is (since it fixes a common situation when optimizing more than once, which with GC gets more common).

Reply:
I would prefer more descriptive names, like `--dae` and `--dae-optimizing`. (This style of `--something` and `--something-optimizing` looks already being used by many passes anyway.) But yeah, they can be follow-ups.